File: ClassicalSharp/Utils/Respawn.cs
// Copyright 2014-2017 ClassicalSharp | Licensed under BSD-3
using System;
using ClassicalSharp.Entities;
using ClassicalSharp.Physics;
using OpenTK;
using BlockID = System.UInt16;

namespace ClassicalSharp {
	
	/// <summary> Shared helper class for respawning an entity. </summary>
	public static class Respawn {

		public static float HighestFreeY(Game game, ref AABB bb) {
			int minX = Utils.Floor(bb.Min.X), maxX = Utils.Floor(bb.Max.X);
			int minY = Utils.Floor(bb.Min.Y), maxY = Utils.Floor(bb.Max.Y);
			int minZ = Utils.Floor(bb.Min.Z), maxZ = Utils.Floor(bb.Max.Z);
			
			float spawnY = float.NegativeInfinity;
			AABB blockBB = default(AABB);
			
			for (int y = minY; y <= maxY; y++)
				for (int z = minZ; z <= maxZ; z++)
					for (int x = minX; x <= maxX; x++)
			{
				BlockID block = game.World.GetPhysicsBlock(x, y, z);
				blockBB.Min = new Vector3(x, y, z) + BlockInfo.MinBB[block];
				blockBB.Max = new Vector3(x, y, z) + BlockInfo.MaxBB[block];
				
				if (BlockInfo.Collide[block] != CollideType.Solid) continue;
				if (!bb.Intersects(blockBB)) continue;				
				spawnY = Math.Max(spawnY, blockBB.Max.Y);
			}
			return spawnY;
		}
		
		public static Vector3 FindSpawnPosition(Game game, float x, float z, Vector3 modelSize) {
			Vector3 spawn = new Vector3(x, 0, z);
			spawn.Y = game.World.Height + Entity.Adjustment;			
			AABB bb = AABB.Make(spawn, modelSize);
			spawn.Y = 0;
			
			for (int y = game.World.Height; y >= 0; y--) {
				float highestY = HighestFreeY(game, ref bb);
				if (highestY != float.NegativeInfinity) {
					spawn.Y = highestY; break;
				}
				bb.Min.Y -= 1; bb.Max.Y -= 1;
			}
			return spawn;
		}
	}
}

File: src/Client/Respawn.c
#if 0
#include "Respawn.h"
#include "ExtMath.h"
#include "World.h"
#include "Block.h"
#include "Funcs.h"
#include "Entity.h"

#define Respawn_NotFound -10000.0f
Real32 Respawn_HighestFreeY(AABB* bb) {
	Int32 minX = Math_Floor(bb->Min.X), maxX = Math_Floor(bb->Max.X);
	Int32 minY = Math_Floor(bb->Min.Y), maxY = Math_Floor(bb->Max.Y);
	Int32 minZ = Math_Floor(bb->Min.Z), maxZ = Math_Floor(bb->Max.Z);

	Real32 spawnY = Respawn_NotFound;
	AABB blockBB;
	Int32 x, y, z;
	Vector3 pos;

	for (y = minY; y <= maxY; y++) {
		pos.Y = (Real32)y;
		for (z = minZ; z <= maxZ; z++) {
			pos.Z = (Real32)z;
			for (x = minX; x <= maxX; x++) {
				pos.X = (Real32)x;
				BlockID block = World_GetPhysicsBlock(x, y, z);
				Vector3_Add(&blockBB.Min, &pos, &Block_MinBB[block]);
				Vector3_Add(&blockBB.Max, &pos, &Block_MaxBB[block]);

				if (Block_Collide[block] != CollideType_Solid) continue;
				if (!AABB_Intersects(bb, &blockBB)) continue;
				spawnY = max(spawnY, blockBB.Max.Y);
			}
		}
	}
	return spawnY;
}

Vector3 Respawn_FindSpawnPosition(Real32 x, Real32 z, Vector3 modelSize) {
	Vector3 spawn = Vector3_Create3(x, 0.0f, z);
	spawn.Y = World_Height + ENTITY_ADJUSTMENT;
	AABB bb;
	AABB_Make(&bb, &spawn, &modelSize);
	spawn.Y = 0.0f;

	Int32 y;
	for (y = World_Height; y >= 0; y--) {
		Real32 highestY = Respawn_HighestFreeY(&bb);
		if (highestY != Respawn_NotFound) {
			spawn.Y = highestY; break;
		}
		bb.Min.Y -= 1; bb.Max.Y -= 1;
	}
	return spawn;
}
#endif